cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -413,19 +416,33 @@ private void addProgram(String phaseName, WorkflowProgramAdder programAdder) {
     for (StageSpec connectorStage : phase.getStagesOfType(Constants.CONNECTOR_TYPE)) {
       phaseConnectorDatasets.put(connectorStage.getName(), connectorDatasets.get(connectorStage.getName()));
     }
-    BatchPhaseSpec batchPhaseSpec = new BatchPhaseSpec(programName, phase, spec.getResources(),
-                                                       spec.getDriverResources(),
-                                                       spec.getClientResources(),
-                                                       spec.isStageLoggingEnabled(),
-                                                       spec.isProcessTimingEnabled(),
-                                                       phaseConnectorDatasets,
-                                                       spec.getNumOfRecordsPreview(),
-                                                       spec.getProperties());
+
+    return new BatchPhaseSpec(programName, phase, spec.getResources(), spec.getDriverResources(),
+                              spec.getClientResources(), spec.isStageLoggingEnabled(), spec.isProcessTimingEnabled(),
+                              phaseConnectorDatasets, spec.getNumOfRecordsPreview(), spec.getProperties());
+  }
+
+  private void addProgram(String phaseName, WorkflowProgramAdder programAdder) {
+    PipelinePhase phase = plan.getPhase(phaseName);
+    // if the origin plan didn't have this name, it means it is a join node
+    // artificially added by the control dag flattening process. So nothing to add, skip it
+    if (phase == null) {
+      return;
+    }
+
+    // can't use phase name as a program name because it might contain invalid characters
+    String programName = "phase-" + phaseNum;
+    phaseNum++;
+
+    BatchPhaseSpec batchPhaseSpec = getPhaseSpec(programName, phase);
 
     Set<String> pluginTypes = batchPhaseSpec.getPhase().getPluginTypes();
     if (pluginTypes.contains(Action.PLUGIN_TYPE)) {
       // actions will be all by themselves in a phase
       programAdder.addAction(new PipelineAction(batchPhaseSpec));
+    } else if (pluginTypes.contains(Condition.PLUGIN_TYPE)) {
+      // conditions will be all by themselves in a phase
+      addCondition(programAdder, phaseName, batchPhaseSpec);
```

Comment:
I think its cleaner if you just do programAdder.condition() here, and move the other logic in addCondition() to the addPrograms() method. That way all the logic related to forking or condition-ing is in the same place